Lib/pathlib/__init__.py

```diff
@@ -584,7 +584,12 @@ def iterdir(self):
         The children are yielded in arbitrary order, and the
         special entries '.' and '..' are not included.
         """
-        return (self._make_child_relpath(name) for name in os.listdir(self))
+        root_dir = str(self)
+        with os.scandir(root_dir) as scandir_it:
+            paths = [entry.path for entry in scandir_it]
```

Comment:
Maybe look into keeping a reference to each `DirEntry` object. Otherwise it's discarding information that makes the `is_dir()`, `is_file()`, `is_symlink()`, and `is_junction()` method calls nearly free when the entry isn't a symlink or when `follow_symlinks` is false.

Reply:
Users expect `stat()` etc to fetch up-to-date information unfortunately, otherwise I'd agree. I've considered adding a new property like `path.status` that would explicitly cache.